src/parser.cpp

```diff
@@ -0,0 +1,637 @@
+                                            "' from units '" + units->getName() + "' is not convertible to decimal number.");
+                        err->setUnits(units);
+                        addError(err);
+                    }
+                } else {
+                    UnitsErrorPtr err = std::make_shared<UnitsError>();
+                    err->setDescription("Invalid attribute '" + attribute->getType() +
+                                        "' found in unit '" + childNode->getAttribute("units") +
+                                        "' from units '" + units->getName() + "'.");
+                    err->setUnits(units);
+                    addError(err);
+                }
+                attribute = attribute->getNext();
+            }
+            units->addUnit(name, prefix, exponent, multiplier, offset);
+        } else {
+            UnitsErrorPtr err = std::make_shared<UnitsError>();
+            err->setDescription("Invalid child element '" + childNode->getType() +
+                                "' found in units '" + units->getName() + "'.");
+            err->setUnits(units);
+            addError(err);
+        }
+        childNode = childNode->getNext();
+    }
+}
+
+void Parser::loadVariable(const VariablePtr &variable, const XmlNodePtr &node)
+{
+    XmlAttributePtr attribute = node->getRootAttribute();
+    while (attribute) {
+        if (attribute->isType("name")) {
+            variable->setName(attribute->getValue());
+        } else if (attribute->isType("units")) {
+            variable->setUnits(attribute->getValue());
+        } else if (attribute->isType("interface")) {
+            variable->setInterfaceType(attribute->getValue());
+        } else if (attribute->isType("initial_value")) {
+            variable->setInitialValue(attribute->getValue());
+        } else {
+            VariableErrorPtr err = std::make_shared<VariableError>();
+            err->setDescription("Invalid attribute '" + attribute->getType() +
+                                "' found in variable '" + node->getAttribute("name") + "'.");
+            err->setVariable(variable);
+            addError(err);
+        }
+        attribute = attribute->getNext();
+    }
+}
+
+void Parser::loadConnection(const ModelPtr &model, const XmlNodePtr &node)
+{
+    // Initialise connection components and variables.
+    ComponentPtr component1 = nullptr;
+    ComponentPtr component2 = nullptr;
+    VariablePtr variable1 = nullptr;
+    VariablePtr variable2 = nullptr;
+    // Load the connection map_components.
+    XmlNodePtr mapComponentsNode = node->getChild();
+    if (!mapComponentsNode) {
+        ModelErrorPtr err = std::make_shared<ModelError>();
+        err->setDescription("Connection does not contain any child elements.");
+        err->setModel(model);
+        addError(err);
+        return;
+    }
+    if (mapComponentsNode->isType("map_components")) {
+        if (mapComponentsNode->hasAttribute("component_1")) {
```

Comment:
What if we have an attribute that is neither `component_1` nor `component_2`? In other words, why not do like what has been done above, i.e. have a `while` loop for attributes?